File: src/test/java/discretemath/bool/expression/forms/ProductOfSumExpressionTest.java
package discretemath.bool.expression.forms;

import java.util.Map;

import discretemath.bool.expression.atomic.BooleanVariable;
import discretemath.bool.expression.atomic.BooleanVariables;
import discretemath.bool.table.FunctionValue;
import discretemath.bool.table.TruthTable;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ProductOfSumExpressionTest {

	@Test
	void evaluate() {
		TruthTable truthTable = TruthTable.forNDegree(3, 1);

		truthTable.setFunctionValue(0, new FunctionValue[]{FunctionValue.TRUE,
														   FunctionValue.FALSE,
														   FunctionValue.FALSE,
														   FunctionValue.TRUE,
														   FunctionValue.TRUE,
														   FunctionValue.TRUE,
														   FunctionValue.FALSE,
														   FunctionValue.FALSE});

		ProductOfSumExpression productOfSumExpression = ProductOfSumExpression.fromTruthTable(truthTable);

		assertEquals("(x+y+z′)(x+y′+z)(x′+y′+z)(x′+y′+z′)", productOfSumExpression.toString());

		assertTrue(productOfSumExpression.evaluate(getXYZValues(false, false, false)));
		assertFalse(productOfSumExpression.evaluate(getXYZValues(false, false, true)));
		assertFalse(productOfSumExpression.evaluate(getXYZValues(false, true, false)));
		assertTrue(productOfSumExpression.evaluate(getXYZValues(false, true, true)));
		assertTrue(productOfSumExpression.evaluate(getXYZValues(true, false, false)));
		assertTrue(productOfSumExpression.evaluate(getXYZValues(true, false, true)));
		assertFalse(productOfSumExpression.evaluate(getXYZValues(true, true, false)));
		assertFalse(productOfSumExpression.evaluate(getXYZValues(true, true, true)));
	}

	private static Map<BooleanVariable, Boolean> getXYZValues(boolean x, boolean y, boolean z) {
		return Map.of(BooleanVariables.X, x, BooleanVariables.Y, y, BooleanVariables.Z, z);
	}
}

File: src/test/scala/discretemath/test/bool/expression/forms/ProductOfSumExpressionTest.scala
package discretemath.test.bool.expression.forms

import discretemath.bool.expression.atomic.BooleanVariables.{X, Y, Z}
import discretemath.bool.expression.forms.ProductOfSumExpression
import discretemath.bool.table.{FunctionValue, TruthTable}
import org.scalatest.funsuite.AnyFunSuite

class ProductOfSumExpressionTest extends AnyFunSuite {

  test("evaluate") {
    val truthTable = TruthTable.forNDegree(3, 1)
    truthTable.setFunctionValue(0, Array[FunctionValue](FunctionValue.TRUE, FunctionValue.FALSE, FunctionValue.FALSE, FunctionValue.TRUE, FunctionValue.TRUE, FunctionValue.TRUE, FunctionValue.FALSE, FunctionValue.FALSE))
    val expression = ProductOfSumExpression(truthTable)
    assert(expression.toString == "(x+y+~z)(x+~y+z)(~x+~y+z)(~x+~y+~z)")
    assert(expression.evaluate(Map(X -> false, Y -> false, Z -> false)))
    assert(!expression.evaluate(Map(X -> false, Y -> false, Z -> true)))
    assert(!expression.evaluate(Map(X -> false, Y -> true, Z -> false)))
    assert(expression.evaluate(Map(X -> false, Y -> true, Z -> true)))
    assert(expression.evaluate(Map(X -> true, Y -> false, Z -> false)))
    assert(expression.evaluate(Map(X -> true, Y -> false, Z -> true)))
    assert(!expression.evaluate(Map(X -> true, Y -> true, Z -> false)))
    assert(!expression.evaluate(Map(X -> true, Y -> true, Z -> true)))
  }
}
